### ISSUE TITLE
Segfault when using -listen and X509Vnc

### ISSUE THREAD

**Issue by exikyut** [TigerVNC/tigervnc#1440]
[Applies to: current GIT head as of around the same time this was filed]

TL;DR, when using `-listen` and an X509Vnc security configuration, a segfault that I believe is unavoidable is encountered during successful session negotiation.

The crash is occurring at

https://github.com/TigerVNC/tigervnc/blob/d80a3d2fa81207c34a31f56073bc4b67101e3c72/common/rfb/CSecurityTLS.cxx#L299-L302

because `client->getServerName()` is `NULL` (the `strlen()` is what explodes). That's the symptom/effect of the crash, not the cause.

I have no idea how to fix this structurally speaking; IMHO the fix feels like it might loosely depend a bit on aesthetic preference, and thus be best informed/implemented by a) whoever has a good mental model of the code's history and b) whoever has to maintain it :) (since, y'know, personal preference and resonance and all that).

Here's a view of the bug coming at the problem/context from the outside in. I'm still learning how to effectively write up this sort of thing, so if the following presentation is fragmented and hard to follow, it might be interesting to just scan it rapidly, or perhaps read it backward.

Firstly I understand that `-listen ...` is not a compound construction of (param, arg), but rather that **a)** `-listen` sets `listenMode` (off in `Configuration::...`), **b)** `vncServerName` is (re)<sup><sub>(ab)</sub></sup>used to store the listening address, **c)** the `listenMode` options are parsed, and **d)** the logic continues/falls-through to `mainloop()`:

https://github.com/TigerVNC/tigervnc/blob/d80a3d2fa81207c34a31f56073bc4b67101e3c72/vncviewer/vncviewer.cxx#L739-L740

https://github.com/TigerVNC/tigervnc/blob/d80a3d2fa81207c34a31f56073bc4b67101e3c72/vncviewer/vncviewer.cxx#L780

https://github.com/TigerVNC/tigervnc/blob/d80a3d2fa81207c34a31f56073bc4b67101e3c72/vncviewer/vncviewer.cxx#L814-L817

https://github.com/TigerVNC/tigervnc/blob/d80a3d2fa81207c34a31f56073bc4b67101e3c72/vncviewer/vncviewer.cxx#L841

https://github.com/TigerVNC/tigervnc/blob/d80a3d2fa81207c34a31f56073bc4b67101e3c72/vncviewer/vncviewer.cxx#L844

`mainloop()` creates a new `CConn()` with the server name and `sock` of the received connection (and on a side note I think is entirely unrelated to this bug, I note the `CConn()` is passed the global `vncServerName`, as opposed to the `vncserver` in `mainloop()`'s prototype which does not seem to be being used anywhere):

https://github.com/TigerVNC/tigervnc/blob/d80a3d2fa81207c34a31f56073bc4b67101e3c72/vncviewer/vncviewer.cxx#L181-L188

Jumping into CConn(), I first make note of `serverHost(0)` (L79) and `sock = socket` (L84):

https://github.com/TigerVNC/tigervnc/blob/d80a3d2fa81207c34a31f56073bc4b67101e3c72/vncviewer/CConn.cxx#L78-L84

Next (just focusing on the Unix side because **a)** that's what I'm testing and **b)** the Unix code is helpfully explicit), I note how `serverHost` is set on L102:

https://github.com/TigerVNC/tigervnc/blob/d80a3d2fa81207c34a31f56073bc4b67101e3c72/vncviewer/CConn.cxx#L97-L105

Finally, the leading `}` below closes the `if (sock == NULL) {` above. There is no `sock != NULL` `else` clause.

https://github.com/TigerVNC/tigervnc/blob/d80a3d2fa81207c34a31f56073bc4b67101e3c72/vncviewer/CConn.cxx#L118-L122
**Thus, when in `listenMode`, `sock != NULL`, `serverHost` is never changed from its default of `serverHost(0)` at the top of `CConn()`, and `setServerName()` is thus set to `NULL`**.

You can kind of see why **a)** I'm like "nooope" \**runs for the hills*\*, **b)** why I zoomed all the way out to talking about options and arguments, and **c)** why I think (I might be wrong) that this is a vaguely semantic/aesthetic problem.

Adding an `else` clause to `CConn()`'s `if (sock == NULL)` would mean that the VNC viewer would describe its own listen address as the server address, breaking TLS CN verification and leading to infinite "Hostname (...) does not match the server certificate" popups forever.

But that's not even mentioning the violation of the separation-of-concerns boundary... which is a good question actually, given its scope of use, *should* a `CConn` know whether it describes a reverse connection?

This is the exact point I begin wandering aimlessly outside of my circle of competence. I'm certainly interested to see how this gets resolved.

I've temporarily added an `} else { serverHost = rfb::strDup(""); }` (yes that required a bit of headscratching :) ) after the logic under `if (sock == NULL) {` in `CConn` for now as a giant hack to allow my use case to work. However, this does nothing to address the fact that in this mode the server is not considering the client name (and I'm not even sure how to make it do that).

Thanks very much for TigerVNC!

<sub>I filed this issue by directly loading `/issues/new`, and only just learned about the templates while filing another issue. My apologies for not using the template.</sub>

**Comment by CendioOssman** [comment]
Yeah, using reverse connections and the X.509 security types is a bit of an odd corner case so I'm not surprised this was overlooked.

It is also a bit unclear how this should work as X.509 needs to have a user specified server name to properly secure things. I suppose the best we can do for now is to always treat it as a server name mismatch.

**Comment by exikyut** [comment]
I actually just realized I initially confused my machines' IP addresses for a minute, and that `sock->getPeerAddress()` can actually do interesting and useful things here.

If `CConn` is changed such that

```
  if(sock == NULL) {
    ...
  } else {
    serverHost = sock->getPeerAddress();
  }
```

then I immediately get a popup reporting that <the inbound/connecting machine's IP> doesn't match the certificate. So it figures it out correctly!

Obviously that doesn't take into account the `#ifndef WIN32`s nearby and only works for Unix.

I was incidentally going to file a feature request about exactly the hostname popup as my next move - thanks very much for the feedback, I was able to fold it into that issue: https://github.com/TigerVNC/tigervnc/issues/1441

**Comment by CendioOssman** [comment]
Unfortunately that isn't a proper way forward. What we want to verify is that the server is the machine that the user expected. So we need some input from the user in order to determine what is expected. We cannot do this automatically.

**Comment by CendioOssman** [comment]
Crash fixed in 7cba36627c9ec5de62d5e29d3ff2d77b216137b0. Thanks for reporting!

Changing the verification of the certificate can be discussed under #1441.
